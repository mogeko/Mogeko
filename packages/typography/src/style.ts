import type { CSSObject } from "@emotion/serialize";

const round = (num: number) => {
  return num
    .toFixed(7)
    .replace(/(\.[0-9]+?)0+$/, "$1")
    .replace(/\.0$/, "");
};
const rem = (px: number) => `${round(px / 16)}rem`;
const em = (px: number, base: number) => `${round(px / base)}em`;
const hsl = (val: string) => `hsl(${val})`;

export const styles: Styles = {
  DEFAULT: {
    a: {
      "--tw-shadow": "inset 0 -1px 0 0",
      "--tw-shadow-colored": "inset 0 -1px 0 0 var(--tw-shadow-color)",
      boxShadow:
        "var(--tw-ring-offset-shadow, 0 0 #0000), var(--tw-ring-shadow, 0 0 #0000), var(--tw-shadow)",
      position: "relative",
      "&:hover": {
        "--tw-shadow": `inset 0 ${em(-8, 16)} 0 0 var(--tw-shadow-color)`,
        "--tw-shadow-color": hsl("var(--primary)"),
        "--tw-shadow-colored": `inset 0 ${em(-8, 16)} 0 0 var(--tw-shadow-color)`,
      },
    },
    blockquote: {
      borderColor: hsl("var(--primary)"),
      borderLeftWidth: "5px",
      fontStyle: "italic",
      lineHeight: rem(28),
      margin: `0 ${rem(28)} ${rem(28)} ${rem(24)}`,
      opacity: 0.8,
      paddingLeft: rem(20),
    },
    code: {
      fontFamily: `ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace`,
      "&:not(:where(.astro-code *))": {
        backgroundColor: hsl("var(--muted)"),
        borderRadius: rem(4),
        fontSize: rem(14),
        fontWeight: 600,
        lineHeight: rem(20),
        padding: `${rem(3.2)} ${rem(4.8)}`,
        position: "relative",
      },
    },
    "h1, h2, h3, h4, h5, h6": {
      letterSpacing: em(-0.9, 36),
      scrollMargin: rem(80),
    },
    h1: {
      fontSize: rem(36),
      fontWeight: 900,
      lineHeight: rem(40),
      "&:not(:first-child)": {
        margin: `${rem(56)} 0 ${rem(28)}`,
      },
    },
    h2: {
      fontSize: rem(30),
      fontWeight: 800,
      lineHeight: rem(36),
      paddingBottom: rem(8),
      "&:not(:first-child)": {
        margin: `${rem(48)} 0 ${rem(24)}`,
      },
    },
    h3: {
      fontSize: rem(24),
      fontWeight: 700,
      lineHeight: rem(32),
      "&:not(:first-child)": {
        margin: `${rem(40)} 0 ${rem(20)}`,
      },
    },
    h4: {
      fontSize: rem(20),
      fontWeight: 600,
      lineHeight: rem(28),
      "&:not(:first-child)": {
        margin: `${rem(32)} 0 ${rem(16)}`,
      },
    },
    img: {
      borderRadius: "calc(var(--radius) - 4px)",
    },
    "ol, ul": {
      fontWeight: 500,
      lineHeight: 1.1,
      marginTop: rem(16),
      "& > li": {
        marginBottom: rem(16),
        paddingLeft: rem(12),
        "&::marker": {
          color: hsl("var(--primary)"),
        },
      },
    },
    ol: {
      listStyleType: "decimal",
    },
    p: {
      letterSpacing: em(-0.16, 16),
      lineHeight: rem(28),
      "&:not(:last-child)": {
        marginTop: rem(24),
      },
    },
    ul: {
      listStyleType: "disc",
    },
  },
  lg: {
    h1: {
      fontSize: rem(48),
      lineHeight: 1,
    },
    h2: {
      fontSize: rem(36),
      lineHeight: rem(40),
    },
    h3: {
      fontSize: rem(30),
      lineHeight: rem(36),
    },
  },
};

export type Styles = {
  [key: string]: CSSObject;
  DEFAULT: CSSObject;
};
