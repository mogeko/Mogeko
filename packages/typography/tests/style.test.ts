import { styles } from "@/style";
import { expect, it, describe } from "vitest";

describe("typography styles", () => {
  it("default styles", () => {
    expect(styles.DEFAULT).toMatchSnapshot();
    expect(styles.lg).toMatchSnapshot();
  });
});
