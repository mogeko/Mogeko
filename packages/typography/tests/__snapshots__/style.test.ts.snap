// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`typography styles > default styles 1`] = `
{
  "--tw-prose-muted": "#e2e8f0",
  "--tw-prose-primary": "#0f172a",
  "--tw-prose-radius": "0.5rem",
  "a": {
    "&:hover": {
      "--tw-shadow": "inset 0 -0.5em 0 0 var(--tw-shadow-color)",
      "--tw-shadow-color": "var(--tw-prose-primary)",
      "--tw-shadow-colored": "inset 0 -0.5em 0 0 var(--tw-shadow-color)",
    },
    "--tw-shadow": "inset 0 -1px 0 0",
    "--tw-shadow-colored": "inset 0 -1px 0 0 var(--tw-shadow-color)",
    "boxShadow": "var(--tw-ring-offset-shadow, 0 0 #0000), var(--tw-ring-shadow, 0 0 #0000), var(--tw-shadow)",
    "position": "relative",
  },
  "blockquote": {
    "borderColor": "var(--tw-prose-primary)",
    "borderLeftWidth": "5px",
    "fontStyle": "italic",
    "lineHeight": "1.75rem",
    "margin": "0 1.75rem 1.75rem 1.5rem",
    "opacity": 0.8,
    "paddingLeft": "1.25rem",
  },
  "code": {
    "&:not(:where(.astro-code *))": {
      "backgroundColor": "var(--tw-prose-muted)",
      "borderRadius": "0.25rem",
      "fontSize": "0.875rem",
      "fontWeight": 600,
      "lineHeight": "1.25rem",
      "padding": "0.2rem 0.3rem",
      "position": "relative",
    },
    "fontFamily": "ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace",
  },
  "h1": {
    "&:not(:first-child)": {
      "margin": "3.5rem 0 1.75rem",
    },
    "fontSize": "2.25rem",
    "fontWeight": 900,
    "lineHeight": "2.5rem",
  },
  "h1, h2, h3, h4, h5, h6": {
    "letterSpacing": "-0.025em",
    "scrollMargin": "5rem",
  },
  "h2": {
    "&:not(:first-child)": {
      "margin": "3rem 0 1.5rem",
    },
    "fontSize": "1.875rem",
    "fontWeight": 800,
    "lineHeight": "2.25rem",
    "paddingBottom": "0.5rem",
  },
  "h3": {
    "&:not(:first-child)": {
      "margin": "2.5rem 0 1.25rem",
    },
    "fontSize": "1.5rem",
    "fontWeight": 700,
    "lineHeight": "2rem",
  },
  "h4": {
    "&:not(:first-child)": {
      "margin": "2rem 0 1rem",
    },
    "fontSize": "1.25rem",
    "fontWeight": 600,
    "lineHeight": "1.75rem",
  },
  "img": {
    "borderRadius": "calc(var(--tw-prose-radius) - 4px)",
  },
  "ol": {
    "listStyleType": "decimal",
  },
  "ol, ul": {
    "& > li": {
      "&::marker": {
        "color": "var(--tw-prose-primary)",
      },
      "marginBottom": "1rem",
      "paddingLeft": "0.75rem",
    },
    "fontWeight": 500,
    "lineHeight": 1.1,
    "marginTop": "1rem",
  },
  "p": {
    "&:not(:last-child)": {
      "marginTop": "1.5rem",
    },
    "letterSpacing": "-0.01em",
    "lineHeight": "1.75rem",
  },
  "ul": {
    "listStyleType": "disc",
  },
}
`;

exports[`typography styles > default styles 2`] = `
{
  "h1": {
    "fontSize": "3rem",
    "lineHeight": 1,
  },
  "h2": {
    "fontSize": "2.25rem",
    "lineHeight": "2.5rem",
  },
  "h3": {
    "fontSize": "1.875rem",
    "lineHeight": "2.25rem",
  },
}
`;
